docs: getting started documentation for web FirebaseUI v7 (#1340)
firebase/firebaseui-web: GETTING_STARTED.md

# Easily add sign-in to your web app with FirebaseUI for Web

FirebaseUI for Web is a set of libraries built on the [Firebase Authentication](https://firebase.google.com/docs/auth) JavaScript SDK. Like the classic `firebaseui` package, it helps you ship authentication flows quickly, but the latest version is a complete rewrite with a modern modular architecture.

FirebaseUI for Web now provides these benefits:

- Modern modular SDK support with `initializeApp(...)` and the current Firebase JS SDK.
- Composable screens, forms, and buttons instead of a single monolithic widget.
- Support for React, Shadcn, and Angular.
- Configurable behaviors for redirect vs popup flows, Google One Tap, anonymous upgrade, phone settings, and more.
- Localization support via `@firebase-oss/ui-translations`.

> - Built-in support for email/password, email link, phone auth, OAuth providers, and multi-factor flows.

Reviewer: Instead of comparing to the old package, this should just list features that would be useful to someone that has never used FirebaseUI before. I think it will end up being a fusion of this list and the list that's currently in the docs

> If you are migrating from the old `firebaseui` package, read [MIGRATION.md](MIGRATION.md) alongside this guide.

Reviewer: ```suggestion
Note: If you are migrating from FirebaseUI v6 or earlier, read [the migration guide](https://github.com/firebase/firebaseui-web/blob/main/MIGRATION.md) in the FirebaseUI repository.
```

> This document follows the same overall flow as the [classic FirebaseUI Web documentation](https://firebase.google.com/docs/auth/web/firebaseui), but updated for the current version of FirebaseUI for Web.

Reviewer: remove

## Before you begin

1. Add Firebase to your web app and enable **Authentication** in the [Firebase console](https://console.firebase.google.com/).
2. Use the modular Firebase JS SDK:

   ```ts
   import { initializeApp } from 'firebase/app';

   const app = initializeApp({
     /* your Firebase config */
   });
   ```

3. Install `firebase` if it is not already in your project:

   ```bash
   npm install firebase
   ```

> 

Reviewer: these can be sub-steps of step 1. They're also out of order (need to install Firebase first)

4. Choose your platform:
   - `@firebase-oss/ui-react` for React apps
   - the FirebaseUI Shadcn registry for Shadcn-based apps

>    - `@firebase-oss/ui-angular` with AngularFire for Angular apps

Reviewer: ```suggestion
   - the FirebaseUI Shadcn registry for [shadcn/ui](https://ui.shadcn.com/)-based React apps
   - `@firebase-oss/ui-angular` with AngularFire for Angular apps
   - `@firebase-oss/ui-react` for React apps without shadcn/ui
```

Reviewer: Let's reorder to shadcn-first

> > The new FirebaseUI for Web does not use the old `firebaseui.auth.AuthUI(firebase.auth())` widget model. Instead, you initialize a shared UI instance with `initializeUI(...)`, then render framework-specific components.

Reviewer: this can be removed

> ## Install packages

Reviewer: same here, reorder to shadcn-first

### React

```bash
npm install @firebase-oss/ui-react@beta @firebase-oss/ui-styles
```

### Shadcn

> Shadcn uses the same React runtime as the React package, but the UI components come from the Firebase registry instead of direct package imports.

Reviewer: ```suggestion
The Firebase shadcn/ui registry provides a convenient way to install FirebaseUI in your React app.
```

> Add the Firebase registry to `components.json`:

Reviewer: ```suggestion
Add the Firebase registry to [`components.json`](https://ui.shadcn.com/docs/components-json):
```

```json
{
  "registries": {
    "@firebase": "https://firebaseopensource.com/r/{name}.json"
  }
}
```

> Then add the components you want to use:

Reviewer: how do I find out which components are available?

Reviewer: oh, maybe we could say that components are listed out in later steps

Author: looks like this now: https://github.com/firebase/firebaseui-web/blob/getting-started-doc/GETTING_STARTED.md?plain=1#L40-L72

```bash
npx shadcn@latest add @firebase/sign-in-auth-screen
```

This installs the underlying React FirebaseUI dependencies for you.

### Angular

```bash
npm install @angular/fire @firebase-oss/ui-angular@beta @firebase-oss/ui-core@beta @firebase-oss/ui-styles@beta
```

## Initialize FirebaseUI

> The old library used a widget instance and `ui.start(...)`. The new library uses a shared UI store created with `initializeUI(...)`.

Reviewer: Let's remove all references to the old library, except in the migration guide blurb [above](https://github.com/firebase/firebaseui-web/pull/1340/changes#r3052368852)

import { initializeUI } from '@firebase-oss/ui-core';
import { FirebaseUIProvider } from '@firebase-oss/ui-react';

const app = initializeApp({
  /* your Firebase config */
});

const ui = initializeUI({
  app,
});

export function AppProviders({ children }: { children: React.ReactNode }) {
  return <FirebaseUIProvider ui={ui}>{children}</FirebaseUIProvider>;
}
```

### Angular

```ts
import { type ApplicationConfig } from '@angular/core';
import { provideFirebaseApp, initializeApp } from '@angular/fire/app';
import { provideFirebaseUI } from '@firebase-oss/ui-angular';
import { initializeUI } from '@firebase-oss/ui-core';

export const appConfig: ApplicationConfig = {
  providers: [
    provideFirebaseApp(() =>
      initializeApp({
        /* your Firebase config */
      }),
    ),
    provideFirebaseUI((apps) =>
      initializeUI({
        app: apps[0],
      }),
    ),
  ],
};
```

## Include styles

If you are using the default React or Angular components, include the FirebaseUI styles:

```css
@import '@firebase-oss/ui-styles/dist.min.css';
```

If you are using Tailwind:

```css
@import 'tailwindcss';
@import '@firebase-oss/ui-styles/tailwind';
```

> If you are using Shadcn, you typically do **not** import FirebaseUI's bundled CSS. The generated components inherit your Shadcn design system instead.

Reviewer: since there's the big caveat for shadcn, maybe this should just be part of the "initialize" code samples.

Author: part of initialize docs now: https://github.com/firebase/firebaseui-web/blob/getting-started-doc/GETTING_STARTED.md?plain=1#L100

## Set up sign-in methods

Before users can sign in, enable each provider you want in **Authentication** -> **Sign-in method** in the Firebase console.

The biggest change from the old library is this:

- Old FirebaseUI used a single `signInOptions` array passed to a widget config.
- New FirebaseUI uses screens, forms, and buttons that you render directly.

> - Cross-cutting configuration now lives in `behaviors` passed to `initializeUI(...)`.

Reviewer: remove

### Email address and password

1. Enable **Email/Password** in the Firebase console.
2. Render `SignInAuthScreen` or `SignUpAuthScreen` in React/Shadcn, or `fui-sign-in-auth-screen` / `fui-sign-up-auth-screen` in Angular.

Optional: require a display name during sign-up:

```ts
import { initializeUI, requireDisplayName } from '@firebase-oss/ui-core';

const ui = initializeUI({
  app,
  behaviors: [requireDisplayName()],
});
```

### Email link authentication

1. Enable **Email/Password** and **Email link (passwordless sign-in)** in the Firebase console.

> 2. Render the email link components from your platform package.

Reviewer: as you mention the specific components in "Email address and password", let's do that for every method

Author: done: https://github.com/firebase/firebaseui-web/blob/getting-started-doc/GETTING_STARTED.md?plain=1#L169-L196

3. Complete sign-in with the current URL using the core helpers when needed.

```ts
import { completeEmailLinkSignIn } from '@firebase-oss/ui-core';

await completeEmailLinkSignIn(ui, window.location.href);
```

### OAuth providers

FirebaseUI for Web supports built-in buttons for providers such as Google, Apple, Facebook, GitHub, Microsoft, and X/Twitter.

1. Enable the provider in the Firebase console.
2. Add your app domain to **Authorized domains** where required.
3. Render the provider buttons you want.

### Phone number

1. Enable **Phone** in the Firebase console.
2. Add your app domain to **Authorized domains**.
3. Render the phone auth screen or form for your platform.

Optional: configure allowed countries, default country, or reCAPTCHA behavior:

```ts
import {
  countryCodes,
  initializeUI,
  recaptchaVerification,
} from '@firebase-oss/ui-core';

const ui = initializeUI({
  app,
  behaviors: [
    countryCodes({
      allowedCountries: ['GB', 'US', 'FR'],
      defaultCountry: 'GB',
    }),
    recaptchaVerification({
      size: 'compact',
      theme: 'light',
    }),
  ],
});
```

## Sign in

> Instead of calling `ui.start('#container', config)`, render the auth screen you want and handle success in component callbacks or Angular outputs.

Reviewer: ```suggestion
Render the auth screen you want and handle success in component callbacks or Angular outputs.
```

    oneTapSignIn({
      clientId: 'YOUR_GOOGLE_WEB_CLIENT_ID',
      autoSelect: false,
      cancelOnTapOutside: false,
    }),
  ],
});
```

Make sure Google sign-in is enabled in the Firebase console, then copy the web client ID from the Google provider settings.

## Terms of service and privacy policy

The old library used `tosUrl` and `privacyPolicyUrl` in the widget config. The new library attaches policy links through the platform provider configuration.

### React

```tsx
import { FirebaseUIProvider } from '@firebase-oss/ui-react';

<FirebaseUIProvider
  ui={ui}
  policies={{
    termsOfServiceUrl: 'https://example.com/terms',
    privacyPolicyUrl: 'https://example.com/privacy',
  }}
>
  {children}
</FirebaseUIProvider>;
```

### Shadcn

Use the same `FirebaseUIProvider` configuration as React.

### Angular

```ts
import { type ApplicationConfig } from '@angular/core';
import { provideFirebaseUIPolicies } from '@firebase-oss/ui-angular';

export const appConfig: ApplicationConfig = {
  providers: [
    provideFirebaseUIPolicies(() => ({
      termsOfServiceUrl: 'https://example.com/terms',
      privacyPolicyUrl: 'https://example.com/privacy',
    })),
  ],
};
```

> ## Upgrading anonymous users

Reviewer: I'm wondering if this should go under a new "Customize behavior" heading, along with the require username behavior

Author: moved it to under customize behavior header: https://github.com/firebase/firebaseui-web/blob/getting-started-doc/GETTING_STARTED.md?plain=1#L222-L259

> The old library supported anonymous account upgrade with `autoUpgradeAnonymousUsers` plus `signInFailure` merge handling. The new library keeps the capability, but it is now configured as a behavior.

Reviewer: remove reference to the old library

```ts
import {
  autoUpgradeAnonymousUsers,
  initializeUI,
} from '@firebase-oss/ui-core';

const ui = initializeUI({
  app,
  behaviors: [
    autoUpgradeAnonymousUsers({
      async onUpgrade(ui, oldUserId, credential) {
        // Migrate or merge user data here if needed.
      },
    }),
  ],
});
```

If you previously handled merge conflicts in v6 callbacks, see [MIGRATION.md](MIGRATION.md) for the updated behavior-based model.

## Translations

> FirebaseUI for Web supports localization through `@firebase-oss/ui-translations`. At the moment, a couple of locales are bundled out of the box, including English (`en-US`) and Czech (`cs-CZ`). You can also register and use your own locale overrides today, and PRs to add more built-in languages are welcomed.

Reviewer: interesting, was Czech a community contribution?

maybe we can have Gemini generate a bunch of different languages

Author: Yes! It was a user contributon. I've opened an issue for adding additional languages: https://github.com/firebase/firebaseui-web/issues/1348

Shadcn uses the same setup as React, because it also uses `@firebase-oss/ui-react` under the hood:

```tsx
import { initializeApp } from 'firebase/app';
import { initializeUI } from '@firebase-oss/ui-core';
import { FirebaseUIProvider } from '@firebase-oss/ui-react';

const app = initializeApp({
  /* your Firebase config */
});

const ui = initializeUI({
  app,
});

export function AppProviders({ children }: { children: React.ReactNode }) {
  return <FirebaseUIProvider ui={ui}>{children}</FirebaseUIProvider>;
}
```

These components use your existing shadcn styles; you typically do not import FirebaseUI's bundled CSS when using the shadcn registry.

### Angular

```ts
import { type ApplicationConfig } from '@angular/core';
import { provideFirebaseApp, initializeApp } from '@angular/fire/app';
import { provideFirebaseUI } from '@firebase-oss/ui-angular';
import { initializeUI } from '@firebase-oss/ui-core';

export const appConfig: ApplicationConfig = {
  providers: [
    provideFirebaseApp(() =>
      initializeApp({
        /* your Firebase config */
      }),
    ),
    provideFirebaseUI((apps) =>
      initializeUI({
        app: apps[0],
      }),
    ),
  ],
};
```

> ## Include styles

Reviewer 2: This section looks different to the [google doc.](https://docs.google.com/document/d/1NQEEjLTbkGMN8NoBUqNGELiKtCP0GmU9U0K-Z9KDcs8/edit?tab=t.0)

> If you are using the default React or Angular components, include the FirebaseUI styles:

Reviewer 2: ```suggestion
FirebaseUI has pre-built UI components for React and Angular. To use the components, you need to include their CSS:

If you are using Tailwind with React or Angular:

```

```css
@import '@firebase-oss/ui-styles/dist.min.css';

> ```

Reviewer 2: ```suggestion
```css
@import 'tailwindcss';
@import '@firebase-oss/ui-styles/tailwind';
```
```

> Otherwise, if you are using Tailwind:

Reviewer 2: ```suggestion
Otherwise, include the FirebaseUI styles:
```

```css
@import 'tailwindcss';

> @import '@firebase-oss/ui-styles/tailwind';

Reviewer 2: ```suggestion
@import '@firebase-oss/ui-styles/dist.min.css';
```
